vendor/sigs.k8s.io/kustomize/api/types/kustomization.go

```diff
@@ -88,6 +95,7 @@ type Kustomization struct {
 	// specification. This can also be done with a patch.
 	Replicas []Replica `json:"replicas,omitempty" yaml:"replicas,omitempty"`
 
+	// Deprecated: Vars will be removed in future release. Migrate to Replacements instead.
```

Comment:
removal of supported functionality from existing versioned kustomize API files is unexpected - kubectl is expected to continue working with existing manifests on upgrade

Reply:
We were worried that backwards-incompatible changes would not be accepted when we updated kustomize in kubectl, so @knverey and I brought it up in several SIG-CLI meetings before deciding to release kustomize v5 and received the message that because kustomize is independently maintained and released, we were free to make breaking changes accordingly as long as we communicate exactly what the changes are.

Anything marked Deprecated still works but has a warning now. But there are a few functionalities that have been changed (so not backwards compatible) that we describe in https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv5.0.0. The ones that will surface in `kubectl kustomize` are:

- https://github.com/kubernetes-sigs/kustomize/pull/4911: Drop support for a very old, legacy style of patches. patches used to be allowed to be used as an alias for patchesStrategicMerge in kustomize v3. You now have to use patchesStrategicMerge explicitly, or update to the new syntax supported by patches. See examples in the PR description of https://github.com/kubernetes-sigs/kustomize/pull/4911.
- https://github.com/kubernetes-sigs/kustomize/issues/4731: Remove a potential build-time side-effect in ConfigMapGenerator and SecretGenerator, which loaded values from the local environment under some circumstances, breaking kustomize build's side-effect-free promise. While this behavior was never intended, we deprecated it and are announcing it as a breaking change since it existed for a long time. See also the Eschewed Features documentation.
- https://github.com/kubernetes-sigs/kustomize/pull/4985: If you previously included .git in an AWS or Azure URL, we will no longer automatically remove that suffix. You may need to add an extra / to replace the .git for the URL to properly resolve.
- https://github.com/kubernetes-sigs/kustomize/pull/4954: Drop support for using gh: as a host (e.g. gh:kubernetes-sigs/kustomize). We were unable to find any usage of or basis for this and believe it may have been targeting a custom gitconfig shorthand syntax.

Reply:
> removal of supported functionality from existing versioned kustomize API files is unexpected

Since we're commenting on a field deprecation line, I just want to make sure it is clear that have NOT nor will we ever remove the field from the v1beta1 Kustomization type. The inline comment is unfortunately not clear to that effect, but the release notes and docs pages are. For example, for vars:

> This field will never be removed from the kustomize.config.k8s.io/v1beta1 Kustomization API, but it will not be included in the kustomize.config.k8s.io/v1 Kustomization API. When Kustomization v1 is available, we will announce the deprecation of the v1beta1 version. There will be at least two releases between deprecation and removal of Kustomization v1beta1 support from the kustomize CLI, and removal itself will happen in a future major version bump.

No work on Kustomization v1 is even scheduled at this point, but we have always assumed we will need follow the API deprecation policy with respect to _kubectl_ releases (kustomize standalone releases are historically much more frequent). Given that the replacement fields are already available (and indeed `kustomize edit fix` will even perform the conversion automatically in most cases), we felt that warning users to switch as much ahead of time as possible would be helpful (and get us more feedback on any blockers), rather than relying on them seeing replacement notices in docs/issues.

Reply:
@liggitt To be clear, we are not actually removing any of the supported kustomization fields. Do you have any specific objections to the behavioral changes we did make (which I described above)?

Reply:
did we sweep for impact of https://github.com/kubernetes-sigs/kustomize/pull/4911 and https://github.com/kubernetes-sigs/kustomize/pull/4985 in publicly visible kustomization.yaml files?

https://github.com/kubernetes-sigs/kustomize/issues/4731 sounds like a bugfix

https://github.com/kubernetes-sigs/kustomize/pull/4954 sounds like it was obscure / unused as far as we could tell

Reply:
it's hard to understand from the release note which talks about breaking changes whether the old `patches` field no longer works in this version or if you get a warning and an auto-conversion to `patchesStrategicMerge`

if we emit a warning and upconvert, that seems ok. if we just dropped support in this release with no prior warning (from kubectl) that seems less ok.

Reply:
Re: `patches`, we _used_ to upconvert `patches` to `patchesStrategicMerge`. Now we don't. There's been a warning printed since kustomize v4.5.6 - that if a user uses `patches` or `patchesStrategicMerge`, they are asked to update to the new `patches` syntax. 

Re: https://github.com/kubernetes-sigs/kustomize/pull/4985 I don't believe we did a sweep (@KnVerey can confirm?) but we at least did research to consider the potential regressions, and from the PR description I believe this is a net bug fix.

Reply:
> did we sweep for impact of [... ] in publicly visible kustomization.yaml files?

For the `.git` suffix change, I'm not sure how we could pull that search off effectively. If you check out the change summary in the [PR description](https://github.com/kubernetes-sigs/kustomize/pull/4985), perhaps the impact is more clear. In the release notes we're trying to err on the side of over-informing to make sure folks that are affected will not be surprised. But what this one amounts to is that Kustomize was previously mutating some AWS/Azure URLs in a way that may have caused invalid inputs to be "fixed", and the fact that it did so was breaking other URLs that _are_ valid.

For the `patches` change I don't think that would be possible to search for (more on this below).

> https://github.com/kubernetes-sigs/kustomize/issues/4731 sounds like a bugfix

That is how we feel as well. Listing it as a breaking change is again us trying to be super aware of the ways changes we make might affect our users and make sure they aren't surprised. We know from issue comments that some folks did discover and find "uses" for this bug in particular, so we even went as far as emitting an execution-time warning to affected users for the past couple releases (included in the last two kubectl versions), "deprecating" the buggy behaviour. So it shouldn't surprise anyone at this point.

> https://github.com/kubernetes-sigs/kustomize/pull/4954 sounds like it was obscure / unused as far as we could tell

Yeah that was our conclusion / reason for removing it (we did do a public repo search for this one).

> it's hard to understand from the release note which talks about breaking changes whether the old patches field no longer works in this version or if you get a warning and an auto-conversion to patchesStrategicMerge

So, the history of the `patches` field we're talking about here is rather unfortunate from an API evolution perspective. This all predates my personal involvement, but IIRC from when we made the removal decision, it originally had the type `[]string`, and the string values could either be file paths or inline SMPs. Next, it was renamed to `patchesStrategicMerge`, which still exists and still has those same semantics. I'm not sure whether the `patches` field itself was briefly been removed at that point or not, but in any case I bet it would have been in a v3-series version that didn't ever go into kubectl. At any rate, it was soon after re-introduced with different syntax and semantics, i.e. [the ones it has today](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/patches/). At some point along the line, 'compatibility' code was introduced that regex scanned the user's Kustomization file before loading it, guessed whether the old syntax was being used, and mutated it if so. That code was hurting users by producing really bad/misleading error messages in some cases where they were trying to use the current syntax and had made a mistake. The old syntax hasn't been documented or valid since v3 at least but possibly even v2. What we did was remove that regex scanning. Maybe Natasha knows differently, but I don't think there was a warning in this case.

Reply:
> Since we're commenting on a field deprecation line, I just want to make sure it is clear that have NOT nor will we ever remove the field from the v1beta1 Kustomization type.

Yeah... "Vars will be removed in future release" sounds ... like it will be removed in a future release :-)

> There's been a warning printed since kustomize v4.5.6

That's helpful to know, at least the previous versions of kubectl would have warned about this.

I actually looked back and the `patches` string field was already marked as deprecated in the first version of kustomize added to kubectl, which is good.

> IIRC from when we made the removal decision, it originally had the type []string, and the string values could either be file paths or inline SMPs. Next, it was renamed to patchesStrategicMerge, which still exists and still has those same semantics. I'm not sure whether the patches field itself was briefly been removed at that point or not, but in any case I bet it would have been in a v3-series version that didn't ever go into kubectl. At any rate, it was soon after re-introduced with different syntax and semantics, i.e. [the ones it has today](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/patches/).

That's is... rather unfortunate. Hopefully there are tests in place now (like the cross-version compat tests Kubernetes has at https://github.com/kubernetes/api/tree/master/testdata) to keep that sort of evolution from occurring again in the future.

Reply:
Thanks for pointing out that suite! We don't have anything like that, so I opened an issue about it: https://github.com/kubernetes-sigs/kustomize/issues/5092